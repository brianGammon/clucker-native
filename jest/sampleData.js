const sampleData1 = {
  userSettings: {
    key: 'user1',
    data: {
      currentFlockId: 'flock1',
    },
  },
  flocks: {
    data: {
      flock1: {
        name: 'Test Flock 1',
        ownedBy: 'user1',
      },
      flock2: {
        name: 'Test Flock 2',
        ownedBy: 'user2',
      },
    },
  },
  chickens: {
    data: {
      chicken1: {
        name: 'The First Chicken',
      },
      chicken2: {
        name: 'The Second Chicken',
      },
    },
  },
  eggs: {
    data: {
      egg1: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-01',
        modified: '2018-01-12T18:05:54.806Z',
        notes:
          'This was the record setting egg from May 2017, real egg was 66.1g on this day',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg2: {
        chickenId: 'chicken2',
        chickenName: 'Baby',
        damaged: false,
        date: '2017-09-24',
        modified: '2017-10-12T18:05:54.806Z',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg3: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-02',
        modified: '2018-08-02T18:05:54.806Z',
        notes: '',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 68.1,
      },
    },
  },
};

export default sampleData1;
