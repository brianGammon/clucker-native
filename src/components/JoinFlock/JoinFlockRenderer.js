import React from 'react';
import {
  Text, View, Button, TextInput,
} from 'react-native';
import styles from './styles';

type Props = {
  flockId: string,
  error: string,
  handleJoinFlock: () => void,
  handleChangeText: (text: string) => void,
};

const JoinFlockRenderer = ({
  flockId,
  error,
  handleJoinFlock,
  handleChangeText,
}: Props) => (
  <View style={{ marginTop: 20 }}>
    <Text style={styles.label}>Join Flock</Text>
    <View style={styles.rowContainer}>
      <TextInput
        style={styles.input}
        value={flockId}
        onChangeText={text => handleChangeText('joinForm', text)}
      />
      <Button title="Join" onPress={handleJoinFlock} />
    </View>
    {error !== '' && <Text>{error}</Text>}
  </View>
);

export default JoinFlockRenderer;
