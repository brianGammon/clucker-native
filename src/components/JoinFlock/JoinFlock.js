/* @flow */
import * as React from 'react';
import { connect } from 'react-redux';
import JoinFlockRenderer from './JoinFlockRenderer';
import { actionTypes } from '../../redux/constants';

type Props = {
  userId: string,
  joinFlock: (userId: string, flockId: string) => void,
  error: string,
  inProgress: boolean,
};

type State = {
  value: string,
  error: string | null,
  touched: boolean,
};

const initialState = {
  value: '',
  error: null,
  touched: false,
};

class JoinFlock extends React.Component<Props, State> {
  state = initialState;

  componentDidUpdate(prevProps) {
    const { inProgress, error } = this.props;
    if (!prevProps.error && error) {
      this.setError(error);
    }
    if (prevProps.inProgress && !inProgress && !error) {
      this.resetForm();
    }
  }

  setError = (error: string) => this.setState({ error, touched: true });

  resetForm = () => {
    this.setState(initialState);
  };

  handleChangeText = (text: string) => {
    this.setState({ value: text, error: null, touched: true });
  };

  handleJoinFlock = () => {
    const { userId, joinFlock } = this.props;
    const { value } = this.state;
    if (!value || value.trim() === '') {
      return this.setError('Please enter a flock ID.');
    }
    return joinFlock(userId, value);
  };

  render() {
    const { userId } = this.props;
    return (
      <JoinFlockRenderer
        userId={userId}
        handleJoinFlock={this.handleJoinFlock}
        handleChangeText={this.handleChangeText}
        {...this.state}
      />
    );
  }
}

const mapStateToProps = ({
  auth: { user },
  joinForm: { error, inProgress },
}) => ({
  userId: user ? user.uid : '',
  error,
  inProgress,
});

const mapDispatchToProps = dispatch => ({
  joinFlock: (userId: string, flockId: string) => dispatch({
    type: actionTypes.JOIN_FLOCK_REQUESTED,
    payload: { userId, flockId },
  }),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(JoinFlock);
