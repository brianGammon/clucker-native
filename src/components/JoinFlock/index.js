import JoinFlock from './JoinFlock';

export default JoinFlock;
