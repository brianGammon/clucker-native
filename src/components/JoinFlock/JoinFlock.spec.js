import React from 'react';
import configureStore from 'redux-mock-store';
import sampleData from '../../../jest/test-data/full.2017-10-to-2018-05.json';
import JoinFlock from '.';

describe('JoinFlock component:', () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const store = mockStore(sampleData);
  test('Should render', () => {
    const wrapper = shallow(<JoinFlock store={store} />).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
