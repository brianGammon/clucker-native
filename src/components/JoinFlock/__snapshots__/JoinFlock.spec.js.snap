// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinFlock component: Should render 1`] = `
<JoinFlockRenderer
  error={null}
  handleChangeText={[Function]}
  handleJoinFlock={[Function]}
  touched={false}
  userId="user1"
  value=""
/>
`;
